### PR TITLE
HTCONDOR-1491 vectorize-backfill

### PR DESCRIPTION
**Pull request by JaimeFrey**
<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

File: src/condor_startd.V6/boinc_mgr.cpp [hunk 1]
```diff
@@ -445,10 +441,8 @@ bool
 BOINC_BackfillMgr::walk( BoincSlotMember member_func )
 {
```
Comment (GregThain) [inline src/condor_startd.V6/boinc_mgr.cpp:442]: Does anyone call BOINC_BackfillMgr::walk?  If not, can we just remove it?

File: src/condor_startd.V6/boinc_mgr.cpp [hunk 2]
```diff
@@ -196,12 +196,13 @@ BOINC_BackfillMgr::addSlot( BOINC_BackfillSlot* /*boinc_slot*/ )
 bool
 BOINC_BackfillMgr::rmSlot( int slot_id )
 {
-	if( ! m_slots[slot_id] ) {
+	size_t old_sz = m_slots.size();
+	auto slot_matches = [=](BackfillSlot* bs){bool match = slot_id == bs->getSlotID(); if (match) delete bs; return match;};
+	std::remove_if(begin(m_slots), end(m_slots), slot_matches);
+	if (m_slots.size() == old_sz) {
+		// No slot found/removed
```
Comment (GregThain) [inline src/condor_startd.V6/boinc_mgr.cpp:203]: Note that `std::remove_if` doesn't actually remove, it just shuffle matches to the end of the sequence, and returns an iterator.  I think you want something like

```
auto it = std::remove_if(begin(m_slots), end(m_slots), slot_matches);
if (it == end(m_slots) {
    // no slots found
    return false;
} else {
   m.erase(it, end(it));
   ...
 ```

**Review by GregThain — CHANGES_REQUESTED** [review]
We need the v.erase for this to work, me thinks.

File: src/condor_startd.V6/boinc_mgr.cpp (hunk 2 above)
Comment (JaimeFrey) [reply src/condor_startd.V6/boinc_mgr.cpp:203]: Ah. remove_if() shuffles non-matches up, leaving holes at the end in an unspecified state. Looks like C++20 gives us erase_if(), which does what I was expecting. Your proposed change looks good.

File: src/condor_startd.V6/boinc_mgr.cpp (hunk 1 above)
Comment (JaimeFrey) [reply src/condor_startd.V6/boinc_mgr.cpp:442]: Sounds good to me.

**Review by GregThain — APPROVED** [review]
LGTM.

**Pull request merged**